sled-agent/src/zone_bundle.rs

```diff
@@ -6,42 +6,466 @@
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    Default,
+    Deserialize,
+    Eq,
+    Hash,
+    JsonSchema,
+    PartialEq,
+    Serialize,
+)]
+#[serde(rename_all = "snake_case")]
+#[non_exhaustive]
+pub enum ZoneBundleCause {
+    /// Some other, unspecified reason.
+    #[default]
+    Other,
+    /// A zone bundle taken when a sled agent finds a zone that it does not
+    /// expect to be running.
+    UnexpectedZone,
+    /// An instance zone was terminated.
+    TerminatedInstance,
+    /// Generated in response to an explicit request to the sled agent.
+    ExplicitRequest,
+}
+
+impl PartialOrd for ZoneBundleCause {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for ZoneBundleCause {
+    fn cmp(&self, other: &Self) -> Ordering {
+        use ZoneBundleCause::*;
+        match (self, other) {
+            (Other, Other) => Ordering::Equal,
+            (Other, UnexpectedZone | TerminatedInstance | ExplicitRequest) => {
+                Ordering::Less
+            }
+            (UnexpectedZone, Other) => Ordering::Greater,
+            (UnexpectedZone, UnexpectedZone) => Ordering::Equal,
+            (UnexpectedZone, TerminatedInstance | ExplicitRequest) => {
+                Ordering::Less
+            }
+            (TerminatedInstance, Other | UnexpectedZone) => Ordering::Greater,
+            (TerminatedInstance, TerminatedInstance) => Ordering::Equal,
+            (TerminatedInstance, ExplicitRequest) => Ordering::Less,
+            (ExplicitRequest, ExplicitRequest) => Ordering::Equal,
+            (ExplicitRequest, Other | UnexpectedZone | TerminatedInstance) => {
+                Ordering::Greater
+            }
+        }
+    }
+}
+
+/// Metadata about a zone bundle.
+#[derive(
+    Clone,
+    Debug,
+    Deserialize,
+    Eq,
+    Hash,
+    JsonSchema,
+    Ord,
+    PartialEq,
+    PartialOrd,
+    Serialize,
+)]
+pub struct ZoneBundleMetadata {
+    /// Identifier for this zone bundle
+    pub id: ZoneBundleId,
+    /// The time at which this zone bundle was created.
+    pub time_created: DateTime<Utc>,
+    /// A version number for this zone bundle.
+    pub version: u8,
+    /// The reason or cause a bundle was created.
     pub cause: ZoneBundleCause,
-    /// Extra directories searched for logfiles for the name zone.
+}
+
+impl ZoneBundleMetadata {
+    const VERSION: u8 = 0;
+
+    /// Create a new set of metadata for the provided zone.
+    pub(crate) fn new(zone_name: &str, cause: ZoneBundleCause) -> Self {
+        Self {
+            id: ZoneBundleId {
+                zone_name: zone_name.to_string(),
+                bundle_id: Uuid::new_v4(),
+            },
+            time_created: Utc::now(),
+            version: Self::VERSION,
+            cause,
+        }
+    }
+}
+
+/// A type managing zone bundle creation and automatic cleanup.
+#[derive(Clone)]
+pub struct ZoneBundler {
+    log: Logger,
+    inner: Arc<Mutex<Inner>>,
```

Comment:
Haha, yeah I get all that. I actually _had_ a channel-based mechanism in the previous commit -- I actually like it, though it does mean you end up doing BS like duplicating a function signature in a message enum variant. Your point about debugging is a good one too, though I've found adding DTrace probes helps.

> at a type system level it looks like we could maybe make this Arc<Inner> and then use standard mutexes around inner.cleanup_context and inner.last_cleanup_at, since those are the only fields we modify, but I doubt that would be logically correct (e.g., it would allow multiple concurrent Inner::create() calls to run simultaneously, which is probably wrong?)

I _think_ multiple calls to `Inner::create()` are probably fine. I think in practice, there won't be multiple calls for the _same_ zone, since there's always synchronization around the `RunningZone` argument, inside whoever is _calling_ this function. Though it would be odd, I don't even think it's an error to create multiple bundles for the same zone simultaneously. What we really want to be exclusive is the cleanup, both of creations and other cleanups. But that's the case already with the task. Does that suggest the sync mutex is the right call here?

Reply:
> What we really want to be exclusive is the cleanup, both of creations and other cleanups. But that's the case already with the task. Does that suggest the sync mutex is the right call here?

I don't think it is the case, right? The task runs periodic cleanups, but cleanups can also be triggered by someone calling `zone_bundler.cleanup()`. That runs cleanup directly, so the mutex is preventing such a direct cleanup from running concurrently with an unluckily-timed periodic cleanup.

Reply:
Yeah, I should have been more clear. I meant that the way this is currently structured, where both the task and the direct call to `ZoneBundler::cleanup()` acquire the same lock, each cleanup is synchronized. I think we're saying the same thing, but I worded that poorly earlier.